### PR TITLE
feat: tell the user which platforms matched when failing

### PR DESCRIPTION
**Pull request by mattculler**
- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----
(CRAFT-3107)

### CODE REVIEW

File: craft_application/services/lifecycle.py [hunk 1]
```diff
@@ -425,7 +425,7 @@ def _validate_build_plan(build_plan: list[models.BuildInfo]) -> None:
         raise errors.EmptyBuildPlanError
 
     if len(build_plan) > 1:
-        raise errors.MultipleBuildsError
+        raise errors.MultipleBuildsError(matching_builds=build_plan)
```
Comment (lengau) [inline craft_application/services/lifecycle.py:428]: The code looks good, but please create a unit test that fires this off and checks the error message.

Comment (mattculler) [reply craft_application/services/lifecycle.py:428]: Of course, good call!

File: craft_application/errors.py [hunk 2]
```diff
@@ -141,8 +142,12 @@ def __init__(self) -> None:
...
     """The build plan contains multiple possible builds."""
 
-    def __init__(self) -> None:
-        message = "Multiple builds match the current platform."
+    def __init__(self, matching_builds: list[models.BuildInfo] | None = None) -> None:
+        message = "Multiple builds match the current platform"
+        if not matching_builds:
+            message += "."
+        else:
+            message += ": " + ", ".join([build.platform for build in matching_builds])
```
Comment (mr-cal) [inline craft_application/errors.py:150]: Don't forget you can do `utils.humanize_list([build.platform for build in matching_builds], "and")`

Just a suggestion, I don't have a preference

**Review by mr-cal — APPROVED** [review]

File: craft_application/errors.py (hunk 2 above)
Comment (mattculler) [reply craft_application/errors.py:150]: I wasn't aware of `humanize_list`, going to make use of it

File: tests/unit/services/test_lifecycle.py [hunk 3]
```diff
@@ -825,8 +825,16 @@ def test_no_builds_error(fake_parts_lifecycle):
 @pytest.mark.parametrize("fake_build_plan", [2, 3, 4], indirect=True)
 def test_multiple_builds_error(fake_parts_lifecycle):
     """Build plan contains more than 1 item."""
-    with pytest.raises(errors.MultipleBuildsError):
+    with pytest.raises(errors.MultipleBuildsError) as e:
         fake_parts_lifecycle.run("prime")
+    assert str(e.value) == (
+        "Multiple builds match the current platform: "
```
Comment (tigarmo) [inline tests/unit/services/test_lifecycle.py:831]: can we test against the actual string here? I'd like to see how the end result looks

Comment (lengau) [reply tests/unit/services/test_lifecycle.py:831]: Agreed

Comment (mattculler) [reply tests/unit/services/test_lifecycle.py]: @tigarmo @lengau How's this look? [38312b9](https://github.com/canonical/craft-application/pull/428/commits/38312b981aa43eb9dda9eca39194593ec5df2c6f)

**Review by tigarmo — APPROVED** [review]

**Pull request merged**